File: Lecture 5/FinalCode - Java/app/src/main/java/com/inovaeg/androidtrainingpharos2020/MoviesListActivity.java
package com.inovaeg.androidtrainingpharos2020;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Arrays;

public class MoviesListActivity extends AppCompatActivity {

    private Boolean loadMore = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movies_list);

        //data
        ArrayList<Movie> firstPageMovies = new ArrayList<>();
        firstPageMovies.add(new Movie("Shaun of the dead", R.drawable.poster_shaun_of_the_dead, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        firstPageMovies.add(new Movie("Fight club", R.drawable.poster_fight_club, new ArrayList<>(Arrays.asList("Drama", "Thriller"))));
        firstPageMovies.add(new Movie("Mr. Nobody", R.drawable.poster_mr_nobody, new ArrayList<>(Arrays.asList("Sci-fi", "Drama"))));
        firstPageMovies.add(new Movie("End Game", R.drawable.poster_end_game, new ArrayList<>(Arrays.asList("action", "Sci-fi"))));
        firstPageMovies.add(new Movie("Sherlock Holmes", R.drawable.poster_sherlock_holmes, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        firstPageMovies.add(new Movie("Knives Out", R.drawable.poster_knives_out, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));

        final ArrayList<Movie> secondPageMovies = new ArrayList<>();
        secondPageMovies.add(new Movie("The Dark Knight", R.drawable.poster_the_dark_knight, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        secondPageMovies.add(new Movie("Inception", R.drawable.poster_inception, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        secondPageMovies.add(new Movie("The Matrix ", R.drawable.poster_the_matrix, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        secondPageMovies.add(new Movie("Spirited Away", R.drawable.poster_spirited_away, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        secondPageMovies.add(new Movie("Interstellar", R.drawable.poster_interstellar, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));
        secondPageMovies.add(new Movie("The Lion King", R.drawable.poster_the_lion_king, new ArrayList<>(Arrays.asList("action", "zombie", "comedy"))));


        final MoviesAdapter moviesAdapter = new MoviesAdapter(firstPageMovies);

        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);

        RecyclerView recyclerView = findViewById(R.id.rv_movies);
        recyclerView.setAdapter(moviesAdapter);
        recyclerView.setLayoutManager(linearLayoutManager);


        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (loadMore)
                    if (dy > 0) {
                        int visibleItemCount = linearLayoutManager.getChildCount();
                        int totalItemCount = linearLayoutManager.getItemCount();
                        int pastVisibleItems = linearLayoutManager.findFirstVisibleItemPosition();
                        if (visibleItemCount + pastVisibleItems >= totalItemCount) {
                            moviesAdapter.appendDate(secondPageMovies);
                            loadMore = false;
                        }
                    }
            }
        });

    }
}

File: Lecture 5/FinalCode - Kotlin/app/src/main/java/com/inovaeg/androidtrainingpharos2020/MoviesListActivity.kt
package com.inovaeg.androidtrainingpharos2020

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_movies_list.*
import java.util.*

class MoviesListActivity : AppCompatActivity() {
    private var loadMore = true
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movies_list)

        //data
        val firstPageMovies = ArrayList<Movie>()
        firstPageMovies.add(Movie("Shaun of the dead", R.drawable.poster_shaun_of_the_dead, arrayListOf("action", "zombie", "comedy")))
        firstPageMovies.add(Movie("Fight club", R.drawable.poster_fight_club, arrayListOf("Drama", "Thriller")))
        firstPageMovies.add(Movie("Mr. Nobody", R.drawable.poster_mr_nobody, arrayListOf("Sci-fi", "Drama")))
        firstPageMovies.add(Movie("End Game", R.drawable.poster_end_game, arrayListOf("action", "Sci-fi")))
        firstPageMovies.add(Movie("Sherlock Holmes", R.drawable.poster_sherlock_holmes, arrayListOf("action", "zombie", "comedy")))
        firstPageMovies.add(Movie("Knives Out", R.drawable.poster_knives_out, arrayListOf("action", "zombie", "comedy")))

        val secondPageMovies = ArrayList<Movie>()
        secondPageMovies.add(Movie("The Dark Knight", R.drawable.poster_the_dark_knight, arrayListOf("action", "zombie", "comedy")))
        secondPageMovies.add(Movie("Inception", R.drawable.poster_inception, arrayListOf("action", "zombie", "comedy")))
        secondPageMovies.add(Movie("The Matrix ", R.drawable.poster_the_matrix, arrayListOf("action", "zombie", "comedy")))
        secondPageMovies.add(Movie("Spirited Away", R.drawable.poster_spirited_away, arrayListOf("action", "zombie", "comedy")))
        secondPageMovies.add(Movie("Interstellar", R.drawable.poster_interstellar, arrayListOf("action", "zombie", "comedy")))
        secondPageMovies.add(Movie("The Lion King", R.drawable.poster_the_lion_king, arrayListOf("action", "zombie", "comedy")))

        val moviesAdapter = MoviesAdapter(firstPageMovies)
        val linearLayoutManager = LinearLayoutManager(this)

        rv_movies.apply {
            adapter = moviesAdapter
            layoutManager = linearLayoutManager

            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    if (loadMore)
                        if (dy > 0) {
                            val visibleItemCount = linearLayoutManager.childCount
                            val totalItemCount = linearLayoutManager.itemCount
                            val pastVisibleItems = linearLayoutManager.findFirstVisibleItemPosition()
                            if (visibleItemCount + pastVisibleItems >= totalItemCount) {
                                moviesAdapter.appendDate(secondPageMovies)
                                loadMore = false
                            }
                        }
                }
            })
        }
    }
}
